modules/postprocessing/src/processors/depthoffield.cpp

```diff
@@ -0,0 +1,427 @@
+        }
+    } else {
+        addToAccumulationBuffer(img, cont);
+    }
+
+    if (evalCount_ < maxEvalCount - 1) {
+        // Prepare camera for rendering again
+        if (!approximate_) {
+            nextOutImg_->copyRepresentationsTo(prevOutImg_.get());
+        }
+        moveCamera(camera, maxEvalCount, focusDepth);
+        evalCount_++;
+    } else {
+        // Set output
+        if (approximate_) {
+            synthesizeLightfield(cont);
+        } else {
+            outport_.setData(nextOutImg_);
+        }
+
+        // Reset camera
+        evalCount_ = (ogType_ == SkewedPerspectiveCamera::classIdentifier) ? 0 : maxEvalCount;
+        camera_.setCamera(std::unique_ptr<Camera>(ogCamera_->clone()));
+    }
+}
+
+void DepthOfField::setupRecursion(size2_t dim, int maxEvalCount, std::shared_ptr<const Image> img) {
+    ogCamera_.reset(camera_.get().clone());
+    ogType_ = camera_.get().getClassIdentifier();
+    if (ogType_ != PerspectiveCamera::classIdentifier) {
+        LogWarn("Intended for use with Perspective Camera. Unexpected behavior may follow.");
+    }
+    camera_.setCamera(SkewedPerspectiveCamera::classIdentifier);
+
+    if (approximate_) {
+        // Prepare light field
+        dimLightField_ = size3_t(dim.x, dim.y, viewCountApprox_.get() + simViewCountApprox_.get());
+        lightField_ = std::make_shared<Volume>(dimLightField_, DataFormat<vec4>::get());
+        lightFieldDepth_ = std::make_shared<Volume>(dimLightField_, DataFormat<float>::get());
+        float* lightFieldDepthData = static_cast<float*>(
+            lightFieldDepth_->getEditableRepresentation<VolumeRAM>()->getData());
+        std::fill(lightFieldDepthData,
+                  lightFieldDepthData + dimLightField_.x * dimLightField_.y * dimLightField_.z,
+                  -1.f);
+
+        // Prepare Halton sequence
+        haltonX_ = util::haltonSequence<float>(2, simViewCountApprox_.get());
+        haltonImg_ = std::make_shared<Image>(size2_t(simViewCountApprox_.get(), 1),
+                                             DataFormat<float>::get());
+        float* haltonData = static_cast<float*>(
+            haltonImg_->getColorLayer()->getEditableRepresentation<LayerRAM>()->getData());
+        std::copy(haltonX_.begin(), haltonX_.end(), haltonData);
+    } else {
+        // Prepare accumulation buffer
+        prevOutImg_ = std::make_shared<Image>(dim, DataFormat<vec4>::get());
+        nextOutImg_ = std::make_shared<Image>(dim, DataFormat<vec4>::get());
+
+        // Prepare Halton sequences
+        haltonX_ = util::haltonSequence<float>(2, maxEvalCount);
+        haltonY_ = util::haltonSequence<float>(3, maxEvalCount);
+    }
+
+    // Update focus depth and aperture ranges based on the depths in the scene, ignoring
+    // background points and including the current value.
+    const float* inputDepthData = static_cast<const float*>(
+        img->getRepresentation<ImageRAM>()->getDepthLayerRAM()->getData());
```

Comment:
Much performance would be gained if this computation is done one the gpu with parallel reduce.

Use parallel reduce or minmax_element on the cpu 
https://en.cppreference.com/w/cpp/algorithm/minmax_element

Reply:
Which version of parallel reduce do you mean? I tried to google it but could only find experimental versions? I can't use minmax_element since I want to ignore the value 1.

Reply:
You can pass a lambda for special comparison in std::minmax_element

I think one have to roll your own parallel reduce algorithm for the GPU unfortunately (our of scope here). For OpenCL we have used CLOGS ( https://github.com/bmerry/clogs ). Should be done outside of this module in that case though.

Reply:
Right, but the lambda should only return true or false, indicating whether or not a < b. I'm not sure how I would ignore the value 1 without it affecting either the min or the max value since they are both evaluated using the same lambda function.

Reply:
Ah, then you can use std::accumulate if you want. See here for an example 
https://github.com/inviwo/inviwo/blob/2205ee567b625a65c8aeccb78cd56ea15d5c4e6c/modules/base/include/modules/base/algorithm/dataminmax.h

Reply:
Thanks! Done